### PR TITLE
Remove Azure OpenAI chat and embedding support

### LINKED ISSUE #2932: Remove Azure OpenAI chat and embedding support
## Context

Azure OpenAI cannot be set up in Copilot V4 the way it documents, and the one
route that still works is indistinguishable from using a custom
OpenAI-compatible endpoint.

Azure needs three values beyond the key: the resource instance name, the
deployment id, and a dated API version. The configure dialog has no field for
any of them, so a provider added in V4 carries none. Chat only reaches Azure at
all because the request builder falls back to a pasted base URL, which is
exactly what the custom OpenAI-compatible provider already accepts. Agent mode
never routed Azure at all, so a model configured there cannot run. #2881 is a
4.0.1 report of this, from a user on the opencode backend whose Azure model
could not appear or be added.

Azure has been the most persistently broken provider in this tracker: #98,
#324, #654, #903, #1302, #1336, #1377, #1489, #1709, #2000, #2290, #2504 and
#2881 span three years, and each fix has been to a routing detail that the
custom OpenAI-compatible path does not have.

Embeddings are the exception that has worked, and removing them is the part with
a real cost. It is included deliberately: keeping an embedding provider whose
credentials the settings UI no longer collects leaves a working feature that
nobody can newly configure.

## User experience

Anyone chatting through Azure loses that provider and re-adds the same endpoint
as a custom OpenAI-compatible provider, pasting the same deployment URL and key.
The requests are identical, so nothing about the answers changes.

Anyone indexing a vault with Azure embeddings loses more: their embedding model
is removed, the selection falls back to the default, and the vault has to be
re-indexed on another provider. For a large vault that is a long rebuild and
possibly API spend, and it is the reason this needs a release note rather than a
silent migration.

Stored Azure keys are deleted rather than left behind. Copilot Plus models
served from Azure upstream are unaffected: they are Copilot-hosted and do not
use this provider.

## Proposed behavior

- [ ] Azure OpenAI no longer appears as a chat provider, and its models are
      absent from the model picker.
- [ ] The Azure embedding model no longer appears in the embedding model list.
- [ ] A vault whose selected embedding model was Azure falls back to the default
      embedding model instead of holding a selection that cannot resolve.
- [ ] An existing Azure configuration, chat or embedding, is removed on upgrade
      along with its stored key, rather than left to fail when used.
- [ ] No other provider changes. In particular, filtering Azure out of the model
      catalog must not reclassify its models as OpenAI-compatible.
- [ ] Documentation stops listing Azure as supported and says what to use
      instead.

## Explicit non-goals

- Repairing Azure by adding the missing settings fields.
- Copilot Plus models whose upstream happens to be Azure. They are Copilot
  hosted and share none of this code.
- Preserving an existing vault index built on Azure embeddings. There is no way
  to re-point embeddings at a different provider without re-embedding.
- Migrating an Azure chat provider automatically into a custom
  OpenAI-compatible one. The endpoint and key are the user's to re-enter.
  Suggested follow-up: Offer a one-click convert for removed providers that map
  onto a custom OpenAI-compatible endpoint.

## Success criteria

- A user with no Azure configuration finds no Azure entry among chat providers
  or embedding models.
- A user upgrading with an Azure chat provider saved finds it gone, with no
  error on launch and every other provider still working.
- A user upgrading with Azure selected as their embedding model finds the
  default embedding model selected instead, and can re-index without editing
  any file by hand.
- Adding a different provider from the catalog still works, and no provider that
  previously resolved as Azure now appears as OpenAI-compatible.
- `npm test` passes with the Azure suites removed and the new migration covered.

## Open questions / Risks

- The stored Azure key is keychain-backed and reachable only through the
  provider row. The migration has to delete the entry before dropping the row,
  or the credential outlives every reference to it.
- Removing the embedding provider changes which model a vault indexes with. The
  migration must clear the selection rather than leave it dangling, since a
  stale embedding selection is the difference between a rebuildable index and a
  confusing failure.
- #2881 asks for Azure to work rather than to be removed. It should be closed
  with an explanation pointing its reporter at the custom OpenAI-compatible
  provider, not silently.

### PR DESCRIPTION
**Pull request by zeroliu**
Fixes logancyang/obsidian-copilot#2932

## Why

Azure OpenAI cannot be set up in Copilot V4 the way it needs to be. It requires
a resource instance name, a deployment id and a dated API version beyond the
key, and the configure dialog has no field for any of them, so a provider added
in V4 carries none of the three.

Chat reaches Azure at all only because the request builder falls back to a
pasted base URL. That is the same thing the custom OpenAI-compatible provider
does with the same URL and the same key, so the Azure entry buys nothing and
costs a settings surface that cannot be completed. Pressing **Test** does not
reveal the gap either, and Agent Mode never routed Azure, so a model configured
there cannot run at all. #2881 is a 4.0.1 report of exactly this, from a user on
the opencode backend whose Azure model would neither appear nor be addable.

Embeddings are the half that has worked. They go anyway: keeping an embedding
provider whose credentials the settings UI no longer collects leaves a feature
that works for whoever already had it and nobody else can newly configure.

## What

Azure disappears as a chat provider and as an embedding provider. Anyone
pointing at an Azure deployment re-adds it as a custom OpenAI-compatible
provider with the same URL and key, and the requests are identical.

| Condition | Before | After |
| --- | --- | --- |
| Browsing chat providers | Azure listed, 157 catalog models behind it | Not listed |
| Configuring an Azure provider | No field for instance, deployment or API version | Not applicable |
| Chatting through a saved Azure provider | Works only when a full deployment URL was pasted | Provider removed; re-add as custom OpenAI-compatible |
| Azure in Agent Mode | Listed but unroutable, so never runs | Not listed |
| Azure as the embedding model | Indexes the vault | Removed, selection falls back to the default, vault needs re-indexing |
| Stored Azure key | Kept | Deleted from the keychain |
| Copilot Plus models served from Azure upstream | Work | Unchanged, they share none of this code |

> **Before:** a user upgrades with Azure selected for embeddings, and indexing
> fails with `No embedding model found for: azure-openai|azure openai`.
>
> **After:** the selection is repointed at the default embedding model on
> upgrade, and indexing runs once the vault is re-indexed.

## Non goal

- Repairing Azure by adding the missing settings fields.
- Copilot Plus models whose upstream happens to be Azure. They are
  Copilot-hosted and reach none of this code.
- Preserving an existing vault index built on Azure embeddings. Embeddings from
  one model are not comparable with another's, so there is nothing to carry over.
- Migrating a saved Azure chat provider automatically into a custom
  OpenAI-compatible one. The endpoint and key are the user's to re-enter.
  Suggested follow-up: Offer a one-click convert for removed providers that map
  onto a custom OpenAI-compatible endpoint.
- Updating documentation. No user-facing doc listed Azure as a provider, so the
  issue's documentation criterion turned out to have nothing to change.

## Screenshot

Not captured. The change is the absence of entries in two lists, and reaching
them means building the plugin and driving the Obsidian desktop app, which this
session cannot do without the user granting application access.

The equivalent evidence is deterministic and in the diff: both pickers are fed
by lists this change filters, and `catalogTransform.test.ts` asserts that a wire
payload carrying `azure`, `azure-cognitive-services` and `openai` transforms to
`["openai"]` alone.

## Risk

**High**

| Criterion | Status | Reason |
| --- | :---: | --- |
| No behavior change, or a cosmetic/copy/docs/config change visible where it renders, or deterministic tests cover the changed behavior | ✅ | 13 tests cover the catalog filter, every slice the migration touches including both spellings of the embedding provider, and the v12 version gate |
| A defect would fail CI or be obvious on first use | ❌ | A migration bug deletes settings silently, and a missed embedding selection surfaces only when the user next indexes |
| A revert fully restores prior state, including persisted data | ❌ | The migration cascades away provider rows, models and keychain entries, deletes the pre-BYOK top-level key, repoints the embedding selection, and stamps `settingsVersion: 12`. A revert restores none of it, and a vault re-indexed on another provider cannot be un-re-indexed |
| No auth, permissions, secrets, or input-handling surface changes | ❌ | Deletes each removed row's keychain entry through the cascade, plus the pre-BYOK top-level `azureOpenAIApiKey` entry |
| No public API, plugin API, message, or on-disk contract changes | ❌ | Removes five `azureOpenAIApi*` settings, an `EmbeddingModelProviders` member, and `"azure"` from the persisted `ProviderType` union |
| No core-path concurrency, async-lifecycle, or state-machine changes | ✅ | The planner is pure and the executor is synchronous |
| No hot-path behavior lacks deterministic coverage | ✅ | Chat construction and embedding construction each lose a branch; the remaining providers are covered by existing suites |
| No new dependency | ✅ | Dependency manifests unchanged |
| Human-only behavior stays in one feature area and surfaces quickly | ❌ | It spans two: the chat picker shows the change immediately, but the embedding half is only visible at the next index |

Review: read `planAzureRemoval` and `executeAzureRemoval` in
`src/settings/migrations/azureRemovalMigration.ts` line by line against
`planBedrockRemoval` and `executeBedrockRemoval` in
`src/settings/migrations/bedrockRemovalMigration.ts`, which they deliberately
mirror. Check that the plan names provider rows rather than traversing what the
cascade owns, and that `referencesRemovedProvider` covers both the
space-separated and underscore spellings, since only the latter appears on old
embedding rows. Then read `UNROUTABLE_PROVIDER_NPM` and
`isUnroutableProvider` in `src/modelManagement/catalog/catalogTransform.ts` and
confirm the drop keys on the SDK package rather than the catalog id, and that it
happens before `transformProvider` rather than in `mapNpmToProviderType`, whose
`default` branch answers `openai-compatible`. Run Verification, including steps
4 through 6.

## Verification

1. On the base branch, add an Azure provider from Settings → Models (BYOK) with
   a pasted deployment URL and a key, add one model, and confirm chat answers.
2. Still on the base branch, select the Azure embedding model in Settings and
   index a small vault so an index exists.
3. Switch to this branch, rebuild, and reload the plugin.
4. Open Settings → Models (BYOK). The Azure provider and its model are gone,
   every other provider still has its key, and searching Add provider for
   "azure" matches nothing.
5. Check the embedding model setting. It reads the default model rather than
   Azure, and re-indexing the vault completes without editing any file by hand.
6. Reload once more. Nothing about Azure reappears, and no error surfaces on
   load or on first search.

## Note

The migration deliberately mirrors the Bedrock one rather than sharing code
with it. Both name their provider rows and hand them to
`ModelManagementCoordinator.removeProvider`, which already owns dropping the
backend enrollments, the configured models, the row and its key, so what is left
in each planner is only the legacy `activeModels` rows and the stored
selections. That residue is short and the two differ in it, Azure carrying two
provider spellings and an embedding selection, so a shared helper would abstract
over about forty lines while hiding the difference that matters.

Both also delete a pre-BYOK top-level key. The BYOK migration copied
`azureOpenAIApiKey` onto a provider row without clearing it, and with the field
now gone from the schema nothing else could name its keychain id, so this is its
last chance to be removed.

Removing Azure also emptied `listProviderModels`'s "cannot list" branch: every
remaining provider type is listable, so the `| null` return and its single
caller-side guard had no producer left and went with it.

The embedding half is the part that would have broken a vault if skipped.
`EmbeddingManager.getEmbeddingModel` throws `No embedding model found for:
<key>` when the stored key is not in its map, and there is no fallback, so a
dangling selection fails indexing outright rather than degrading. The row itself
needs no removal: `filterUnsupportedEmbeddingModels` already drops any row whose
provider left `EmbeddingModelProviders`, on every settings write. The stored
selection is the one thing nothing else cleans.

### CODE REVIEW

File: src/modelManagement/catalog/builtinDefinitions.ts [hunk 1]
```diff
@@ -49,11 +49,4 @@ export const CUSTOM_OPENAI_DEFINITION: ProviderDefinition = {
...
   ...LOCAL_PROVIDER_DEFINITIONS,
   CUSTOM_OPENAI_DEFINITION,
-  {
-    id: "azure-openai",
-    displayName: "Azure OpenAI",
-    providerType: "azure",
-    requiresApiKey: true,
-    modelInputHint: "matches your Azure deployment name",
-  },
 ];
```
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/modelManagement/catalog/builtinDefinitions.ts:52]: It's fine to not include a story here.

File: src/LLMProviders/chatModelManager.ts [hunk 2]
```diff
@@ -257,46 +225,6 @@ export default class ChatModelManager {
               },
         }),
       },
```
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/LLMProviders/chatModelManager.ts:196]: stop supporting Azure chat model even if it's saved.

File: src/settings/migrations/azureRemovalMigration.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ * names Azure. Returns `null` when the vault never configured Azure.
+ *
+ * @param settings - Hydrated settings snapshot to plan against.
+ */
+export function planAzureRemoval(settings: CopilotSettings): RemovedProviderPlan | null {
+  const base = planRemovedProviderCleanup(settings, AZURE);
+  const patch: Partial<CopilotSettings> = { ...(base?.patch ?? {}) };
+
+  if (selectsAzureEmbedding(settings.embeddingModelKey)) {
+    patch.embeddingModelKey = DEFAULT_SETTINGS.embeddingModelKey;
```
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/settings/migrations/azureRemovalMigration.ts]: Orama embedding will be removed

File: src/settings/migrations/azureRemovalMigration.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  return { patch, keychainIds: base?.keychainIds ?? [] };
+}
+
+/**
+ * Side-effecting executor. Applies the plan and deletes the stored API keys.
+ *
+ * @param settings - Hydrated settings snapshot the plan is computed from.
+ */
+export function executeAzureRemoval(settings: CopilotSettings): void {
+  applyRemovedProviderPlan(planAzureRemoval(settings), "azure-removal");
```
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/settings/migrations/azureRemovalMigration.ts]: Already in place at HEAD. This comment is anchored to ed26e309, which the rework replaced.

`executeAzureRemoval` calls `deleteLegacyApiKey()` unconditionally, before the plan is computed, so a vault that typed an Azure key but never enabled a model still loses the keychain entry. Covered by the test `deletes the pre-BYOK top-level key even for a vault with nothing else to clean`.

File: src/settings/migrations/removedProviderCleanup.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+/**
+ * Whether a persisted model key names the removed provider. Legacy keys are
+ * `name|provider`, optionally prefixed with an agent backend id, so the
+ * provider is always the trailing segment.
+ *
+ * @param modelKey - Stored selection to test.
+ * @param legacyProvider - Provider value the key would end with.
+ */
+export function keyNamesProvider(modelKey: string | undefined, legacyProvider: string): boolean {
+  return modelKey?.endsWith(`|${legacyProvider}`) ?? false;
```
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/settings/migrations/removedProviderCleanup.ts]: `removedProviderCleanup.ts` no longer exists. This is anchored to ed26e309; the rework deleted that shared module, and each migration now hands its provider rows to `ModelManagementCoordinator.removeProvider`, which already owned the four-step cascade. There is no shared contract left to test in isolation.

File: src/modelManagement/catalog/catalogTransform.ts [hunk 6]
```diff
@@ -30,14 +30,23 @@ export function looksLikeEmbeddingModel(idOrHaystack: string): boolean {
...
- * https://github.com/logancyang/obsidian-copilot/issues/2928
+ * `default` branch answers `openai-compatible`: removing an entry from the
+ * mapping alone would leave its models in the picker behind an endpoint that
+ * cannot answer them.
  */
-const UNROUTABLE_PROVIDER_IDS: ReadonlySet<string> = new Set(["amazon-bedrock"]);
+const UNROUTABLE_PROVIDER_IDS: ReadonlySet<string> = new Set([
+  "amazon-bedrock",
+  "azure",
+  "azure-cognitive-services",
```
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/modelManagement/catalog/catalogTransform.ts]: Fixed in 94500f3, and the catalog data backs the concern harder than the comment does. models.dev already publishes two providers under `@ai-sdk/azure` today, `azure` and `azure-cognitive-services`, so the family case is not hypothetical.

`UNROUTABLE_PROVIDER_NPM` now keys on the npm package rather than the id, which covers the Bedrock family on the same rule. That is also the semantically correct key: `npm` names the client that would have to route the provider, which is exactly what "Copilot cannot route this" means, while an id is a catalog label.

I also dropped the test that defended the id key against a catalog that stops publishing `npm`. All 193 providers in the live catalog publish it, so it guarded a case with no instances.

File: src/settings/migrations/azureRemovalMigration.test.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+          providers: { az: provider("az", "azure", "kc-az"), ant: provider("ant", "anthropic") },
+          configuredModels: [configuredModel("cm-az", "az"), configuredModel("cm-ant", "ant")],
+        })
+      );
+      expect(Object.keys(plan?.patch.providers ?? {})).toEqual(["ant"]);
+      expect(plan?.patch.configuredModels?.map((m) => m.configuredModelId)).toEqual(["cm-ant"]);
+      expect(plan?.keychainIds).toEqual(["kc-az"]);
+    });
+
+    it("repoints an embedding selection that named Azure at the default model", () => {
```
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/settings/migrations/azureRemovalMigration.test.ts]: Fixed in 94500f3. Appended the issue URL to the six `it(...)` descriptions covering the Azure-specific branches: the embedding repoint, its `azure_openai` and no-provider-row edge cases, the legacy chat-model path, and the two executor cases.

The rule also asks for the URL in the nearby code comment, which this comment does not mention. The rationale was sitting in the `planAzureRemoval` JSDoc, so I moved it onto the branch itself with the URL. Doing that caught a wrong symbol in my own comment: it named `EmbeddingManager.getEmbeddingModel`, which does not exist. The method is `getEmbeddingsAPI`.
